<br/>
<span style="font-size:40pt">Cheatsheet!
</span>
<span style="font-size:10pt">😎
</span>
<br/>
###### <span style="color:darkgray">Authors:</span> <span style="color:gray">Adam Plowman</span> <span style="color:darkgray">and</span> <span style="color:gray">Maria Yankova</span>

# The Jupyter Notebook

### How to start a Jupyter notebook
1. Open a terminal
   - In Windows: press the Windows key, then type `cmd` and press `Enter`.
   - On a Mac, type `terminal` in spotlight and press `Enter`.
2. Type `jupyter notebook` and press `Enter`

### Cells
A Jupyter notebook consists of cells.
* **Create a new cell**: click on the plus icon  <img src="https://www.dropbox.com/s/nmctihrfc076dg8/jupyter_newcell.png?dl=1" style="display:inline" /> 
* **Delete a cell**: click on the scissors icon  <img src="https://www.dropbox.com/s/uxdudjxp3xya20a/jupyter_delcell.png?dl=1" style="display:inline" />
* **Edit a cell**: double click on the cell to enter edit mode.
* **Evaluate a cell**: press `Shift` + `Enter`.

There a two different types of cells:
* *Code* cells - in these, you can type programming code (such as Python code)
* Text cells (using something called *Markdown*)

The cell type can be changed between *Code* and *Markdown* using the dropdown box at the top.

#  Python
## 0. Some helpful ideas
**Modules**

Sometimes we need to add extra bits of functionality to Python which aren't available by default. An example we use here is called the Numpy package. To use the Numpy package, we *import* it, like this:

In [1]:
import numpy as np

By typing the `as np` part, we can save ourselves some typing whenever we want to use Numpy later in our code.

** Comments **

Comments are a way of helping you and other people understand your code. Comments are ignored by Python when your code is evaluated and so you can add explanations of what your code is doing in normal language like this:

In [2]:
# here we assign a variable age to the number 10:
age = 10 

We use the `#` symbol to tell Python that the whole line is a comment. Notice also that the Jupyter notebook helpfully changes the colour of comment lines. 

**Indentation**

In Python, we use indentation to organise different sections of code. In the example below, `print('True')` is part of the `if` statement (see section 4.) and so its execution depends on whether `1 < 4` is true. On the other hand, the  `print('Hello')` statement is independent of the `if` statement and always executes.

In [3]:
if 1 < 4:
    print('True')
    
print('Hello')

True
Hello


There are many examples in Python of using a colon `:` followed by an indented block of code. See section 4. (`if` statements), section 5. (`for` loops) and section 6. (Functions). Without indentation, Python will complain! See how this **does not work**:

In [4]:
if 1 < 4:
print('True')

IndentationError: expected an indented block (<ipython-input-4-c6a3a4a6287d>, line 2)

** Getting help **

You can find the documentation for Python functions by typing the function name, followed by a question mark, like this:


In [5]:
print?

If you want to know how to do something in Python, it's always a good idea to search Google.

## 1. Variables
Take a look at this simple Python code:

In [6]:
a = 4

This is called variable assignment. We *assign* the value `4` to the variable named `a`. Practically, variables names are usually more descriptive than a single letter.

Variables don't have to be integer numbers, they can have many different types, such as:
* decimal numbers (known as floats)
* text (known as strings)
* True or False (known as booleans)

Some examples of using different variable types are below. Here, we are describing some information about a person called Bob. We are also showing the type of the variable after assigning a value to it.

In [7]:
bob_name = 'Bob'
type(bob_name)

str

In [8]:
bob_age = 7
type(bob_age)

int

In [9]:
bob_height = 1.5
type(bob_height)

float

In [10]:
bob_blue_eyes = True
type(bob_blue_eyes)

bool

**Testing for equality**

Above, we assigned values to some variables using a single equal sign `=`. 

When we want to check if a variable is equal to something, we use a double equal sign `==` like this:

In [11]:
# Testing for equality with two equal signs
bob_age == 5

False

The result from the equality test is either `True` or `False` depending on whether the condition is met, `False` in this case. Notice how this is different to variable assignment, which uses one equal sign.

In [12]:
# Assigning Anna's age to a variable with a single equal sign
anna_age = 5

## 2. Lists 

A list is another type of variable. Lists are sequences of objects. Here is an example of a list, which we have named `my_list`:

In [13]:
my_list = [2, 4, 6, 8, 10]
type(my_list)

list

Lists are created using square brackets. Commas are used to separate the elements of the list. 

List elements don't have to all be the same type:

In [14]:
my_second_list = ['this is a string!', 9.4, True, 18]

In fact, lists can contain variables which we've previously assigned:

In [17]:
person_info = [bob_name, bob_age, bob_height, bob_blue_eyes]

print(person_info)

['Bob', 7, 1.5, True]


### 2.1 Indexing and slicing
We can access the elements of a list by indexing and slicing. To get the first element of a list we do this:

In [18]:
my_list[0]

2

This is called indexing. Notice how indexing starts at zero and not one! If we want multiple elements, we use slicing. Let's say we want the second, third and fourth elements:

In [19]:
my_list[1:4]

[4, 6, 8]

The `1:4` in the square brackets means get all the elements with indices from `1` (inclusive) to `4` (exclusive). Some more examples of slicing are shown in the table below, where coloured blocks represent the elements we get from each slicing example.
<table>
<tr>
<td rowspan="2" style="padding:1em">`my_list`<br /> element</td>
<td rowspan="2" style="padding:1em">Index</td>
<td colspan="6" style="padding:1em; text-align:center">Slicing examples</td>
</tr>
<tr>
<td style="padding:1em">`my_list[1:4]`</td>
<td style="padding:1em">`my_list[0:1]`</td>
<td style="padding:1em">`my_list[2:]`</td>
<td style="padding:1em">`my_list[:3]`</td>
<td style="padding:1em">`my_list[:-2]`</td>
<td style="padding:1em">`my_list[::2]`</td>
</tr>
<tr>
    <td style="text-align:center">2</td>
    <td style="text-align:center">0</td>
    <td></td>
    <td style="background-color:lightpink"></td>
    <td></td>
    <td style="background-color:lightseagreen"></td>
    <td style="background-color:lightsteelblue"></td>
    <td style="background-color:lightskyblue"></td>
</tr>
<tr>
    <td style="text-align:center">4</td>
    <td style="text-align:center">1</td>
    <td style="background-color:lightblue"></td>
    <td></td>
    <td></td>
    <td style="background-color:lightseagreen"></td>
    <td style="background-color:lightsteelblue"></td>
    <td></td>
</tr>
<tr>
    <td style="text-align:center">6</td>
    <td style="text-align:center">2</td>
    <td style="background-color:lightblue"></td>
    <td></td>
    <td style="background-color:lightsalmon"></td>
    <td style="background-color:lightseagreen"></td>
    <td style="background-color:lightsteelblue"></td>
    <td style="background-color:lightskyblue"></td>
</tr>
<tr>
    <td style="text-align:center">8</td>
    <td style="text-align:center">3</td>
    <td style="background-color:lightblue"></td>
    <td></td>
    <td style="background-color:lightsalmon"></td>
    <td></td>
    <td></td>
    <td></td>    
</tr>
<tr>
    <td style="text-align:center">10</td>
    <td style="text-align:center">4</td>
    <td></td>
    <td></td>
    <td style="background-color:lightsalmon"></td>
    <td></td>
    <td></td>
    <td style="background-color:lightskyblue"></td>    
</tr>
</table>
<br/>
<br/>
Some things to notice about slicing:
* Omitting an index before the colon is the same as using a zero. E.g. `my_list[:3]` is the same as `my_list[0:3]`.
* We can use negative indices after the colon to count back from the end of a list.
* Slices always return a list, so `my_list[0:1]` is a list with a single element at index `0`.

### 2.1 Add elements to a list
To add a new element to the end of a list:

In [20]:
my_list.append('a new element!')
print(my_list)

[2, 4, 6, 8, 10, 'a new element!']


### 2.2 Remove elements from a list
To remove the element with index 3:

In [21]:
my_list.pop(3)
print(my_list)

[2, 4, 6, 10, 'a new element!']


### 2.3 Finding the length of a list

In [22]:
len(my_list)

5

## 3. Arrays
Arrays are like lists, but can be multi-dimensional. A matrix in maths can be considered to be a two-dimensional (2D) array. Here is an example of a 2D array, using the Numpy package (referenced here as `np`) which we previously imported:

In [23]:
my_array = np.array([[1,2,3], [4,5,6], [7,8,9]])
print(my_array)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


### 3.1 Indexing and slicing
Indexing and slicing works for arrays in the same way as for lists. The only difference is that we use commas to separate each dimension we want to index. For instance, we index a 2D array like this: `my_array[row_index, column_index]`, where `row_index` and `column_index` are each just like the indices (or slices) we used for lists. Rows and columns are indexed like this: ![rowcol](https://www.dropbox.com/s/u6n2yi0me2xye2m/xy-rowcol.png?dl=1)

So to get the array element in the first row and third column, we do this:

In [24]:
my_array[0, 2]

3

We can slice arrays just like lists. For instance, to get the first row of the array, the row index would be `0` and the column index would be a slice `0:3` to get all columns.

In [25]:
my_array[0, 0:3]

array([1, 2, 3])

In practice, we can simply use a colon `:` instead of `0:3` to get all elements along a particular dimension:

In [26]:
my_array[0, :]

array([1, 2, 3])

### 3.2 Adding and removing elements from arrays
Once a Numpy array is created, it's impossible to add or remove elements from it. (There are ways to effectively do this, but generally they are inefficient - Google *array concatenation in numpy*.) Therefore, we tend to first create an array which has the correct size, and then later worry about filling it up with data, as seen in the next sub-section.

### 3.3 Creating arrays 

Arrays are always created with a specific size (i.e. number of rows and columns).

**An array filled with given data**

In [27]:
my_data_array = np.array([[5,3,7],[2,7,-1]])
print(my_data_array)

[[ 5  3  7]
 [ 2  7 -1]]


**An array filled with zeros**

This is useful when we know how much data we want to store, but we don't know what it is yet. In this way, we first create an array of zeros and then later populate it when we have the data.

In [28]:
my_zeros_array = np.zeros((4,3))
print(my_zeros_array)

[[ 0.  0.  0.]
 [ 0.  0.  0.]
 [ 0.  0.  0.]
 [ 0.  0.  0.]]


**An array filled with ones**

In [29]:
my_ones_array = np.ones((2, 5))
print(my_ones_array)

[[ 1.  1.  1.  1.  1.]
 [ 1.  1.  1.  1.  1.]]


**Specifying data type**

All elements in a Numpy array must have the same type. This is different to lists, where elements in the same list can be of different types. For instance, a list could have two elements: an integer and a string, whereas the elements in an array must all be integers or all be floats for example.

Let's create an array filled with ones, but let's make sure that the ones are represented as integers, instead of floats which is the default:

In [30]:
my_integer_array = np.ones((2,5), dtype=int)
print(my_integer_array)

[[1 1 1 1 1]
 [1 1 1 1 1]]


If we check the type of the array like we did for other variables we notice that the actual type of the elements isn't revealed:

In [31]:
type(my_integer_array)

numpy.ndarray

`numpy.ndarray` just means that the variable is a Numpy array. To find the type of the array elements we do: 

In [32]:
my_integer_array.dtype

dtype('int64')

### 3.4 Populating arrays

Adding data to an array is simply done like this:

In [33]:
my_zeros_array[1, :] = [1, 2, 3]
print(my_zeros_array)

[[ 0.  0.  0.]
 [ 1.  2.  3.]
 [ 0.  0.  0.]
 [ 0.  0.  0.]]


We assign the data to the elements of the array using slicing and indexing. Notice how the data has to have the correct size. In this example, we are populating the three columns in the second row with three numbers.

### 3.4 Finding the size of an array

In [34]:
np.shape(my_zeros_array)

(4, 3)

## 4. `if` statements
Sometimes we want our code to make a choice. We use `if` statements to evaluate some code only *if* some condition is true. Let's see an example:

In [35]:
# Check whether Bob is tall enough to ride a rollercoaster or not.

if bob_height > 1.3:
    print('Yay! Bob can have fun on the rollercoaster!')

Yay! Bob can have fun on the rollercoaster!


In the above example, we first check if Bob's height is greater than 1.3 using the `if` statement with the greater than symbol `>`. If this is true, then we print some text. We can also evaluate some code if the `if` statement is not true. Here's the example again, but with an `else` statement:

In [36]:
anna_height = 1.2

if anna_height > 1.3:
    print('Yay! You can have fun on the rollercoaster!')
else:
    print('Sorry, you are too short! Maybe next year...')

Sorry, you are too short! Maybe next year...


We can also be more specific by testing multiple conditions using the `elif` statement:

In [37]:
anna_height = 1.2

if anna_height > 1.3:
    print('Yay! You can have fun on the rollercoaster!')
elif 1.1 < anna_height < 1.3:
    print('You are too short for the rollercoaster, but have fun on the Teacups!')
else:
    print('Sorry, you are too short! Maybe next year...')

You are too short for the rollercoaster, but have fun on the Teacups!


## 5. `for` loops

Sometimes we want to repeat the same code multiple times. For example, we might want to do the same thing to each element in a list. First, let's look at a simple example of a `for` loop. Let's print each element in a list:

In [38]:
my_number_list = [0,1,2]
for i in my_number_list:
    print(i)

0
1
2


This `for` loop is identical to the following code:

In [39]:
i = my_number_list[0]
print(i)
i = my_number_list[1]
print(i)
i = my_number_list[2]
print(i)

0
1
2


We can see how the `for` loop is easier to read and type! This is even more true if we want to do something more useful. Let's say we have a list of numbers we want to add together:

In [40]:
num_list = [10,20,30,40,50,60,70]

total = num_list[0] + num_list[1] + num_list[2] + num_list[3] + num_list[4] + num_list[5] + num_list[6]

print (total)

280


If we had a list with 100 elements, imagine how long this would take to type! A better way is to use a `for` loop:

In [41]:
total = 0
for num in num_list:        # num takes the value of each element in num_list: num = num_list[0]; num = num_list[1] ...
    total = total + num     # this value is added to total, so for the first two steps of  the loop:
                            # total = 0 + num_list[0]
                            # total = 0 + num_list[0] + num_list[1] and so on.

print(total) 

280


We can understand how the loop works by looking at the table below. Each step of the loop is called an iteration (shown as rows in the table). So, for each iteration the code `total = total + num` is evaluated. The number of iterations is equal to the number of elements in the list `num_list`.

<table>
<tr>
<td style="padding:0.5em">Iteration </td>
<td style="padding:0.5em">`num`</td>
<td style="padding:0.5em; text-align:center">`total`</td>
</tr>
<tr>
<td style="padding:0.5em">0</td>
<td style="padding:0.5em">10</td>
<td style="padding:0.5em">10</td>
</tr>
<tr>
<td style="padding:0.5em">1</td>
<td style="padding:0.5em">20</td>
<td style="padding:0.5em">30</td>
</tr>
<tr>
<td style="padding:0.5em">2</td>
<td style="padding:0.5em">30</td>
<td style="padding:0.5em">60</td>
</tr>

<tr>
<td style="padding:0.5em">3</td>
<td style="padding:0.5em">40</td>
<td style="padding:0.5em">100</td>
</tr>
<tr>
<td style="padding:0.5em">4</td>
<td style="padding:0.5em">50</td>
<td style="padding:0.5em">150</td>
</tr>
<tr>
<td style="padding:0.5em">5</td>
<td style="padding:0.5em">60</td>
<td style="padding:0.5em">210</td>
</tr>
<tr>
<td style="padding:0.5em">6</td>
<td style="padding:0.5em">70</td>
<td style="padding:0.5em">280</td>
</tr>
</table>

## 6. Functions
In the examples so far, when we wanted to display some text or a variable, we used the code `print()`. This is an example of a Python function.

A function is a block of code that does something specific. We invoke a function (use it) by typing its name followed by parentheses, like this: `function_name()`. Some functions take inputs (these are known as parameters or arguments); these are typed in between the parenthesis, and separated by commas, like this: `function_name(x, y, z)`. In the case of the `print()` function, the input is the text or variable we want to display.

Some functions also produce output which can be assigned to variables. For example, the output from the function for finding the size of an array (see section 3.4) can be assigned to a variable:

In [42]:
array_size = np.shape(my_zeros_array)
print(array_size)

(4, 3)


Here, the function `shape()` is part of the Numpy package, and therefore, we type `np.shape` instead of just `shape()` so Python knows where to find the function. To be clear: `my_zeros_array` is the input argument to the function `shape()`. Python finds the number of rows and columns in the input array and returns the answer, which we assign to a new variable called `array_size`.

`print()` is an example of a Python *built-in* function. We can also define our own functions like this:

In [42]:
def my_func(x):
    return 2 + x

print(my_func(2))

4


Here, we have defined a function called `my_func` which takes a single number as its input, adds two to it and then returns the result as its output. In maths, such a function would be defined like this: $f(x) = 2 + x$.


To create a function, we start with the `def` keyword to indicate we will be defining a new function. Next, we type the name of our new function, followed by its parameters in parentheses and a colon `:`. The *function body* follows on the next lines with indentation. This is the code which is to be executed whenever you invoke the function. In the above example, all we are doing is a simple sum, but more complicated functions may include many lines of code.

** Useful functions **

<table>

<tr>
<td style="padding:0.5em">**Function name** </td>
<td style="padding:0.5em">**Description**</td>
<td style="padding:0.5em;width:35%">**Example**</td>
</tr>

<tr>
<td style="padding:0.5em">`print(my_variable)` </td>
<td style="padding:0.5em">Displays a variable or text on the screen.</td>
<td style="padding:0.5em">
`a = 5 
print(a)`
</td>
</tr>

<tr>
<td style="padding:0.5em">`type(my_variable)` </td>
<td style="padding:0.5em">Returns the type of a variable e.g. `int`, `float`, `str`, `numpy.ndarray`.</td>
<td style="padding:0.5em">
`type(a)`
</td>
</tr>

<tr>
<td style="padding:0.5em">`len(my_object)` </td>
<td style="padding:0.5em">Returns the length (number of elements) of a list or other object.</td>
<td style="padding:0.5em">
`my_list = [1,2,3]
len(my_list)`
</td>
</tr>

<tr>
<td style="padding:0.5em">`my_list.sort()` </td>
<td style="padding:0.5em">Sorts a list in ascending or alphabetiacal order depending on the type. The original list is modified. </td>
<td style="padding:0.5em;">
`names = ['Eve','Jo','Alice']
names.sort()
print(names)`
</td>
</tr>

<tr>
<td style="padding:0.5em">`np.linspace(start,stop,N)` </td>
<td style="padding:0.5em">Returns an array of `N` evenly spaced numbers between `start` and `stop`. The example generates 10 numbers between 10 and 100 (i.e. 10, 20, 30, ..., 100).</td>
<td style="padding:0.5em;">
`np.linspace(10,100,10)`
</td>
</tr>

<tr>
<td style="padding:0.5em">`np.sin(my_variable)` </td>
<td style="padding:0.5em">Returns the sine of `my_variable` (in radians). If `my_variable` is an array, the sine of each element is calculated and returned in a new array. A similar function exists for cosine, called `np.cos(my_variable)`. </td>
<td style="padding:0.5em;">
`my_array = np.array([0.0, 5.0])
np.sin(my_array)`
</td>
</tr>

<tr>
<td style="padding:0.5em">`np.array(my_data)` </td>
<td style="padding:0.5em">Creates an array using `my_data`.</td>
<td style="padding:0.5em;">
`my_array = np.array([0, 1],[3, 4])`
</td>
</tr>

<tr>
<td style="padding:0.5em">`np.shape(my_array)` </td>
<td style="padding:0.5em">Returns the shape (e.g. number of rows and columns) of `my_array`.</td>
<td style="padding:0.5em;">
</td>
</tr>

<tr>
<td style="padding:0.5em">`np.tile(my_array,reps)` </td>
<td style="padding:0.5em">Construct an array by repeating `my_array` the number of times given by `reps`. In the example, the original array 'my_array' is tiled (repeated) three times in the first dimension (rows) and two times in the second dimension (columns). So the original array has a shape of two rows and two columns (2, 2), and the returned array has a shape of six rows and four columns (6, 4).</td>
<td style="padding:0.5em;">
`my_array = np.array([0, 1],[3, 4])
np.tile(my_array,(3,2))`
</td>
</tr>

<tr>
<td style="padding:0.5em">`plt.plot(x,y)` </td>
<td style="padding:0.5em">Displays a line plot of the data given by 'x' and 'y'.</td>
<td style="padding:0.5em;">
`x = [0, 1, 2, 3, 4]
y = [5, 3, 7, 6, 10]
plt.plot(x,y)`
</td>
</tr>

<tr>
<td style="padding:0.5em">
`plt.imshow(my_array)
plt.matshow(my_array)` </td>
<td style="padding:0.5em">Displays an image represented by the values in `my_array`. `my_array` can be two or three dimensional, see documentation (type `plt.imshow?`) for more details.</td>
<td style="padding:0.5em;">
</td>
</tr>



</table>